cluster/src/dunit/scala/org/apache/spark/sql/NorthWindDUnitTest.scala

```diff
@@ -508,128 +532,293 @@ class NorthWindDUnitTest(s: String) extends ClusterManagerTestBase(s) {
+      snappyDF = snappyDF.coalesce(1).orderBy(col1, col: _*)
+      writeToFile(snappyDF, snappyDest, snc)
+      pw.println(s"${queryNum} Result Collected in file $snappyDest")
+    if (sparkFile.listFiles() == null) {
+      sparkDF = sparkDF.coalesce(1).orderBy(col1, col: _*)
+      writeToFile(sparkDF, sparkDest, snc)
+      pw.println(s"${queryNum} Result Collected in file $sparkDest")
+    }
+    val expectedFile = sparkFile.listFiles.filter(_.getName.endsWith(".csv"))
+    val actualFile = snappyFile.listFiles.filter(_.getName.endsWith(".csv"))
+    val expectedLineSet = Source.fromFile(expectedFile.iterator.next()).getLines()
+    val actualLineSet = Source.fromFile(actualFile.iterator.next()).getLines
+    while (expectedLineSet.hasNext && actualLineSet.hasNext) {
+      val expectedLine = expectedLineSet.next()
+      val actualLine = actualLineSet.next()
+      if (!actualLine.equals(expectedLine)) {
+        pw.println(s"\n** For ${queryNum} result mismatch observed**")
+        pw.println(s"\nExpected Result \n: $expectedLine")
+        pw.println(s"\nActual Result   \n: $actualLine")
+        pw.println(s"\nQuery =" + sqlString + " Table Type : " + tableType)
+        assert(false, s"\n** For ${queryNum} result mismatch observed** \nExpected Result \n: $expectedLine \nActual Result   \n: $actualLine \nQuery =" + sqlString + " Table Type : " + tableType)
+      }
+    }
+    if (actualLineSet.hasNext || expectedLineSet.hasNext) {
+      pw.println(s"\nFor ${queryNum} result count mismatch observed")
+      assert(false, s"\nFor ${queryNum} result count mismatch observed")
+    }
+    pw.flush()
+  }
+
+  def assertJoinFullResultSet(snc: SnappyContext, sqlString: String, numRows: Int, queryNum: String, tableType: String, pw: PrintWriter, sqlContext: SQLContext): Any = {
+    snc.sql("set spark.sql.crossJoin.enabled = true")
+    sqlContext.sql("set spark.sql.crossJoin.enabled = true")
+    assertQueryFullResultSet(snc, sqlString, numRows, queryNum, tableType, pw, sqlContext)
+  }
+
+  def dataTypeConverter(row: Row): Row = {
+    val md = row.toSeq.map {
+      //case d: Double => math.floor(d * 1000.0 + 0.5) // round to three digits
+      case d: Double => "%18.1f".format(d).trim().toDouble
+      case de: BigDecimal => {
+        de.setScale(2, BigDecimal.RoundingMode.HALF_UP)
+      }
+      case i: Integer => {
+        i
+      }
+      case v => v
+    }
+    Row.fromSeq(md)
+  }
+
+  def writeToFile(df: DataFrame, dest: String, snc: SnappyContext): Unit = {
+    import snc.implicits._
+    df.map(dataTypeConverter)(RowEncoder(df.schema))
+      .map(row => {
+        var str = ""
+        row.toSeq.foreach(e => {
+          if (e != null)
+            str = str + e.toString + ","
+          else
+            str = str + "NULL" + ","
+        })
+        str
+      }).write.format("org.apache.spark.sql.execution.datasources.csv.CSVFileFormat").option("header", false).save(dest)
```

Comment:
Please avoid CSV writer. From what I have seen it is really slow. All you need in this method is df.collect().mkString(",")